x/zkauth/ante/zkauth.go

```diff
@@ -27,25 +31,123 @@ func (zka ZKAuthMsgDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bo
 		Basically, in the case of zkauth msg, ephPubKey must be idempotent for each msg.
 	*/
 
+	isZKAuthTx, zkMsgs, pubKeys, err := isZKAuthTx(tx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !isZKAuthTx {
+		svd := authante.NewSigVerificationDecorator(zka.ak, zka.signModeHandler)
+		return svd.AnteHandle(ctx, tx, simulate, next)
+	}
+
+	for i, zkMsg := range zkMsgs {
+		// verify ZKAuth signature
+		if err := zkauthtypes.VerifyZKAuthSignature(ctx, zka.zk, pubKeys[i].Bytes(), zkMsg); err != nil {
+			return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "invalid zkauth signature, %s", err)
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+type ZKAuthSetPubKeyDecorator struct {
+	zk zkauthtypes.ZKAuthKeeper
+	ak authante.AccountKeeper
+}
+
+func NewZKAuthSetPubKeyDecorator(zk zkauthtypes.ZKAuthKeeper, ak authante.AccountKeeper) ZKAuthSetPubKeyDecorator {
+	return ZKAuthSetPubKeyDecorator{zk: zk, ak: ak}
+}
+
+func (zsp ZKAuthSetPubKeyDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	isZKAuthTx, zkMsgs, _, err := isZKAuthTx(tx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !isZKAuthTx {
+		spk := authante.NewSetPubKeyDecorator(zsp.ak)
+		return spk.AnteHandle(ctx, tx, simulate, next)
+	}
+
+	for _, zkMsg := range zkMsgs {
+		msgs, err := zkMsg.GetMessages()
+		if err != nil {
+			return ctx, err
+		}
+		for _, msg := range msgs {
+			for _, signer := range msg.GetSigners() {
+				zsp.ak.SetAccount(ctx, zsp.ak.NewAccountWithAddress(ctx, signer))
+			}
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+type ZKAuthIncrementSequenceDecorator struct {
+	ak authante.AccountKeeper
+}
+
+func NewIncrementSequenceDecorator(ak authante.AccountKeeper) ZKAuthIncrementSequenceDecorator {
+	return ZKAuthIncrementSequenceDecorator{
+		ak: ak,
+	}
+}
+
+func (zkisd ZKAuthIncrementSequenceDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	isZKAuthTx, zkMsgs, _, err := isZKAuthTx(tx)
+	if err != nil {
+		return ctx, err
+	}
+
+	if !isZKAuthTx {
+		isd := authante.NewIncrementSequenceDecorator(zkisd.ak)
+		return isd.AnteHandle(ctx, tx, simulate, next)
+	}
+
+	for _, zkMsg := range zkMsgs {
+		msgs, err := zkMsg.GetMessages()
+		if err != nil {
+			return ctx, err
+		}
+		for _, msg := range msgs {
+			for _, signer := range msg.GetSigners() {
+				acc := zkisd.ak.GetAccount(ctx, signer)
+				if err := acc.SetSequence(acc.GetSequence() + 1); err != nil {
+					panic(err)
+				}
+
+				zkisd.ak.SetAccount(ctx, acc)
+			}
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+func isZKAuthTx(tx sdk.Tx) (bool, []*zkauthtypes.MsgExecution, []types.PubKey, error) {
```

Comment:
It appears that isZKAuthTx is responsible for more than just name.
It seems like it can be broken down into three functions for each role.
- isZKAuthTx
- getMsgExecutionFromTx
- getPubkeysFromTx

Reply:
I tried separating it into getMsgExecutionFromTx and getPubkeysFromTx. What do you think?

Reply:
I feel strange that the return value of isZKAuthTx includes pubkey and msgExecution.
The previous suggestion meant processing getMsgExecutionFromTx and getPubkeysFromTx after isZKAuthTx, which returns bool or (bool, error).

I think one way is to rename isZKAuthTx.
e.g) getZKAuthInfoFromTx

Reply:
getMsgExecutionFromTx and getPubkeysFromTx after isZKAuthTx can certainly be called in that order, but I wanted to only call one function with multiple ante handlers, so I grouped them together.